In [2]:
import pandas as pd
import numpy as np
import matplotlib.pyplot as plt
import seaborn as sns

In [3]:
data=pd.read_csv('/content/housing.csv')
data.head()

0.00632  18.00   2.310  0  0.5380  6.5750  65.20  4.0900   1  296.0  15.30 396.90   4.98  24.00
0   0.02731   0.00   7.070  0  0.4690  6.4210  78...                                             
1   0.02729   0.00   7.070  0  0.4690  7.1850  61...                                             
2   0.03237   0.00   2.180  0  0.4580  6.9980  45...                                             
3   0.06905   0.00   2.180  0  0.4580  7.1470  54...                                             
4   0.02985   0.00   2.180  0  0.4580  6.4300  58...

In [5]:
column_names = ['CRIM', 'ZN', 'INDUS', 'CHAS', 'NOX', 'RM', 'AGE', 'DIS', 'RAD', 'TAX', 'PTRATIO', 'B', 'LSTAT', 'Price']
df=pd.read_csv("housing.csv", delimiter=r"\s+", names=column_names)
df.head()

CRIM    ZN  INDUS  CHAS    NOX     RM   AGE     DIS  RAD    TAX  \
0  0.00632  18.0   2.31     0  0.538  6.575  65.2  4.0900    1  296.0   
1  0.02731   0.0   7.07     0  0.469  6.421  78.9  4.9671    2  242.0   
2  0.02729   0.0   7.07     0  0.469  7.185  61.1  4.9671    2  242.0   
3  0.03237   0.0   2.18     0  0.458  6.998  45.8  6.0622    3  222.0   
4  0.06905   0.0   2.18     0  0.458  7.147  54.2  6.0622    3  222.0   

   PTRATIO       B  LSTAT  Price  
0     15.3  396.90   4.98   24.0  
1     17.8  396.90   9.14   21.6  
2     17.8  392.83   4.03   34.7  
3     18.7  394.63   2.94   33.4  
4     18.7  396.90   5.33   36.2

In [6]:
X=df.iloc[:,:-1]
y=df.iloc[:,-1]

In [7]:
X.head()

CRIM    ZN  INDUS  CHAS    NOX     RM   AGE     DIS  RAD    TAX  \
0  0.00632  18.0   2.31     0  0.538  6.575  65.2  4.0900    1  296.0   
1  0.02731   0.0   7.07     0  0.469  6.421  78.9  4.9671    2  242.0   
2  0.02729   0.0   7.07     0  0.469  7.185  61.1  4.9671    2  242.0   
3  0.03237   0.0   2.18     0  0.458  6.998  45.8  6.0622    3  222.0   
4  0.06905   0.0   2.18     0  0.458  7.147  54.2  6.0622    3  222.0   

   PTRATIO       B  LSTAT  
0     15.3  396.90   4.98  
1     17.8  396.90   9.14  
2     17.8  392.83   4.03  
3     18.7  394.63   2.94  
4     18.7  396.90   5.33

In [8]:
y.head()

0    24.0
1    21.6
2    34.7
3    33.4
4    36.2
Name: Price, dtype: float64

<h1>Linear Regression</h1>

In [14]:
from sklearn.linear_model import LinearRegression
lin_reg=LinearRegression()
lin_reg.fit(X,y)

LinearRegression()

In [15]:
from sklearn.model_selection import cross_val_score
scores=cross_val_score(lin_reg,X,y,scoring='neg_mean_squared_error',cv=5)


**sross_val_score**: This function splits the dataset into cv (cross-validation) folds and evaluates the model on each fold<br>
**The Mean Squared Error (MSE)** measures the average squared difference between predicted values (
𝑦
^
𝑖
y
^
​
  
i
​
 ) and actual values (
𝑦
𝑖
y
i).

In [16]:
scores

array([-12.46030057, -26.04862111, -33.07413798, -80.76237112,
       -33.31360656])

In [17]:
mean_score=np.mean(scores)
mean_score

-37.13180746769887

In [18]:
lin_reg.predict(X.iloc[[0]].values.reshape(1,-1))

/usr/local/lib/python3.11/dist-packages/sklearn/utils/validation.py:2739: UserWarning: X does not have valid feature names, but LinearRegression was fitted with feature names
  warnings.warn(


array([30.00384338])

<h1>Ridge Regression</h1>

In [23]:
from sklearn.linear_model import Ridge
ridge_reg=Ridge()
# ridge_reg.fit(X,y)

In [24]:
from sklearn.model_selection import GridSearchCV
params={'alpha':[1e-15,1e-10,1e-8,1e-3,1e-2,1,5,10,15,20]}

In [25]:
ridge_regressor=GridSearchCV(ridge_reg,params,scoring='neg_mean_squared_error',cv=5)

In [26]:
ridge_regressor.fit(X,y)

GridSearchCV(cv=5, estimator=Ridge(),
             param_grid={'alpha': [1e-15, 1e-10, 1e-08, 0.001, 0.01, 1, 5, 10,
                                   15, 20]},
             scoring='neg_mean_squared_error')

In [27]:
ridge_regressor.predict(X.iloc[[0]].values.reshape(1,-1))

/usr/local/lib/python3.11/dist-packages/sklearn/utils/validation.py:2739: UserWarning: X does not have valid feature names, but Ridge was fitted with feature names
  warnings.warn(


array([30.78201924])

In [29]:
print(ridge_regressor.best_params_)
print(ridge_regressor.best_score_)

{'alpha': 20}
-32.380250251825125


**2. GridSearchCV**
GridSearchCV performs an exhaustive search over a specified parameter grid to find the best combination of hyperparameters for a model.<br>
It uses cross-validation to evaluate each combination, ensuring robust performance estimation.<br>

Ridge regression prevents overfitting by shrinking coefficients.
Tuning alpha allows you to control the regularization strength.

**Hyperparameter tuning** is the process of finding the best set of hyperparameters for a machine learning model to optimize its performance on a given task. Unlike model parameters (e.g., weights in a linear regression), hyperparameters are set before training the model and cannot be learned from the data.

**2. Why is Hyperparameter Tuning Important?**
The right hyperparameters improve model performance, reduce overfitting/underfitting, and ensure the model generalizes well to unseen data.
Poorly chosen hyperparameters may lead to:<br>
**Overfitting:** Model performs well on training data but poorly on test data.<br>
**Underfitting:** Model performs poorly on both training and test data.

**<h1>Lasso Regression</h1>**


In [30]:
from sklearn.linear_model import Lasso
from sklearn.model_selection import GridSearchCV
lasso=Lasso()

In [31]:
params={'alpha':[1e-15,1e-10,1e-8,1e-3,1e-2,1,5,10,15,20]}

In [32]:
lasso_regressor=GridSearchCV(lasso,params,scoring='neg_mean_squared_error',cv=5)

In [33]:
lasso_regressor.fit(X,y)

/usr/local/lib/python3.11/dist-packages/sklearn/linear_model/_coordinate_descent.py:695: ConvergenceWarning: Objective did not converge. You might want to increase the number of iterations, check the scale of the features or consider increasing regularisation. Duality gap: 4.633e+03, tolerance: 3.919e+00
  model = cd_fast.enet_coordinate_descent(
/usr/local/lib/python3.11/dist-packages/sklearn/linear_model/_coordinate_descent.py:695: ConvergenceWarning: Objective did not converge. You might want to increase the number of iterations, check the scale of the features or consider increasing regularisation. Duality gap: 3.760e+03, tolerance: 3.307e+00
  model = cd_fast.enet_coordinate_descent(
/usr/local/lib/python3.11/dist-packages/sklearn/linear_model/_coordinate_descent.py:695: ConvergenceWarning: Objective did not converge. You might want to increase the number of iterations, check the scale of the features or consider increasing regularisation. Duality gap: 4.210e+03, tolerance: 2.814e

GridSearchCV(cv=5, estimator=Lasso(),
             param_grid={'alpha': [1e-15, 1e-10, 1e-08, 0.001, 0.01, 1, 5, 10,
                                   15, 20]},
             scoring='neg_mean_squared_error')

In [34]:
print(lasso_regressor.best_params_)
print(lasso_regressor.best_score_)

{'alpha': 1}
-35.53158022069486


In [35]:
lasso_regressor.predict(X.iloc[[0]].values.reshape(1,-1))

/usr/local/lib/python3.11/dist-packages/sklearn/utils/validation.py:2739: UserWarning: X does not have valid feature names, but Lasso was fitted with feature names
  warnings.warn(


array([30.99753918])

A more negative score  indicates a higher MSE, meaning worse model performance.<br>
A less negative score  indicates a lower MSE, meaning better model performance.

**Linear Regression**: Mean score =
−
37.1318<br>
**Ridge Regression:** Best score =
−
32.3803<br>
**Lasso Regression**: Best score =
−
35.5316<br>
Since these are negative mean squared errors (neg-MSE):
<br>
A higher (less negative) value indicates better performance.
Convert the scores to positive MSE for interpretation:<br>
MSE
=
−
neg-MSE

**Which Model is Best?**
Ridge Regression performs the best (
32.38
32.38) as it achieves the lowest MSE.<br>
This suggests that your data likely benefits more from L2 regularization (shrinking coefficients) rather than L1 regularization (eliminating coefficients).